File: src/main/java/com/example/instagramclone/core/aop/annotation/RateLimit.java
package com.example.instagramclone.core.aop.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * [Day 24] Redis INCR + EXPIRE 기반 Rate Limiting 어노테이션.
 *
 * <p>사용 예:
 * <pre>
 *   {@literal @}RateLimit(action = "dm.send", key = "#senderId", limit = 30, windowSeconds = 60)
 *   public DirectMessageResponse sendMessage(Long senderId, DmSendRequest request) {
 *       // 비즈니스 로직
 *   }
 * </pre>
 *
 * <p><b>키 구성</b>: {@code ratelimit:{action}:{key 평가 결과}}
 *
 * <p><b>왜 SpEL 인가?</b><br>
 * - HTTP 요청은 SecurityContextHolder 에 인증 정보가 있지만, STOMP 메시지 핸들러
 *   스레드에는 SecurityContext 가 없다 ({@code Principal} 만 있음).
 * - 결국 어떤 컨텍스트에서 호출되든 "유저를 식별하는 값" 은 항상
 *   <b>메서드 파라미터</b> 로 들어온다 (예: {@code senderId}, {@code loginMemberId}).
 * - 그래서 Day 18 {@code @DistributedLock} 와 똑같이 SpEL 로 파라미터에서 직접 뽑는다.
 *   Aspect 가 SecurityContext / STOMP / IP 같은 호출 컨텍스트에 결합되지 않는다.
 *
 * <p><b>동작</b>: 메서드 호출 직전 {@code INCR} 수행 → 결과가 {@code limit} 초과 시
 * {@link com.example.instagramclone.core.exception.RateLimitException} 을 던져 메서드 실행 자체를 막는다.
 */
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface RateLimit {

    /** 액션 식별자. Redis 키의 중간 세그먼트로 쓰인다. 예: "dm.send", "follow.change" */
    String action();

    /**
     * 카운터 분리 키 — SpEL 표현식. 메서드 파라미터 이름을 {@code #이름} 으로 참조한다.
     * 예: {@code "#senderId"}, {@code "#loginMemberId"}.
     * 비로그인 IP 기반은 호출 측에서 IP 를 인자로 넘겨 같은 패턴을 쓴다.
     */
    String key();

    /** 윈도우 동안 허용되는 최대 요청 수. */
    int limit();

    /** 윈도우 크기(초). 첫 요청 시 이 값으로 TTL 이 설정된다. */
    int windowSeconds();
}

File: src/main/resources/scripts/rate-limit.lua
-- Day 24 과제 1: Rate Limit 원자화 스크립트
--
-- KEYS[1]   : Redis 키 (예: ratelimit:dm.send:42)
-- ARGV[1]   : 윈도우 크기(초)
--
-- 동작:
--   1) INCR 로 카운터를 1 증가시킨다.
--   2) 결과가 1이면 (= 이 스크립트 실행이 이 키의 첫 호출) EXPIRE 로 TTL 을 건다.
--   3) 증가 후 값을 반환한다.
--
-- 원자성:
--   Redis 는 Lua 스크립트를 싱글 스레드로, 한 덩어리로 실행한다.
--   INCR 와 EXPIRE 사이에 다른 클라이언트의 명령이 끼어들 수 없다.
--   → "INCR 했는데 EXPIRE 못 걸고 죽음" 상황 자체가 불가능해진다.

local current = redis.call('INCR', KEYS[1])

if current == 1 then
    redis.call('EXPIRE', KEYS[1], ARGV[1])
end

return current
